### ISSUE TITLE
Gradient accumulation yields worse results than the equivalent batch size

### ISSUE THREAD

**Issue by benjamin-marie** [huggingface/trl#2175]
I expected a training configuration with per_device_train_batch_size=1 and gradient_accumulation_steps=32 to yield the same (or similar) result to per_device_train_batch_size=32 and gradient_accumulation_steps=1 but that's not the case, the former is much worse.
I ran several experiments with SmolLM-135M and Llama 3.2 1B, using always the same seed, and the results are consistent with this observation.

![image](https://github.com/user-attachments/assets/d0d8de5e-469b-4f79-922d-b72a1e7c3788)
![image](https://github.com/user-attachments/assets/452aa89a-737f-477f-bc34-09b744bf8e14)

Maybe I misunderstand something here?

My training code is in [this Colab notebook](https://colab.research.google.com/drive/17g7zVSvGragLiB6DlHGW-D3X0TyfxKHg?usp=sharing). I ran this notebook to draw the learning curves above, restarting the notebook between each training to avoid OOM.
Note that I have the same observations with Qwen2.

**Comment by mayank31398** [comment]
Hey, this is expected behaviour.
FSDP-1 only allows accumulation in 16-bit precision.
This is not the case for FSDP-2 which allows accumulation in both 16-bit and 32-bit.

**Comment by mayank31398** [comment]
documentation for FSDP-1:
<img width="661" alt="Screenshot 2024-10-06 at 2 34 34 AM" src="https://github.com/user-attachments/assets/190b1e60-ea13-4810-a536-4f9ef62e8ae7">
documentation for FSDP-2:
<img width="692" alt="Screenshot 2024-10-06 at 2 35 10 AM" src="https://github.com/user-attachments/assets/4c9e7ee1-0646-4812-9132-ad7ec862a5cd">

**Comment by benjamin-marie** [comment]
Interesting, I didn't know this. But I don't think it matters, I would be surprised that TRL uses FSDP's reduce-scatter for single GPU training.

**Comment by qgallouedec** [comment]
Hi, thanks for reporting this.
Can you share your system info and the code you use for training?

**Comment by benjamin-marie** [comment]
Sure, it's all in the notebook I linked to in my first post. I ran this notebook on Colab with the A100.

**Comment by teknium1** [comment]
Someone tried in fp32 and it didnt help so it doesnt seem  to be the reason 

https://x.com/bnjmn_marie/status/1842464802636980564

**Comment by vigneshwaran** [comment]
Have you tried full/mixed precision AdamW optimiser?

**Comment by benjamin-marie** [comment]
Yes:

![image](https://github.com/user-attachments/assets/d25e7b20-7551-47fe-b758-01f750636738)

This configuration uses fp32 and adamw_torch.

**Comment by danielhanchen** [comment]
I'm writing up a report about this - I think I managed to fix it :)
(Yes it is in fact a subtle bug!) - will tweet and post about it in like 8 - 10 hours!

**Comment by shimmyshimmer** [comment]
We have fixed the issue guys!

Tweet: https://twitter.com/UnslothAI/status/1846231235749990699
Blogpost: https://unsloth.ai/blog/gradient

**Comment by geronimi73** [comment]
> We have fixed the issue guys!

nice! feel like fixing it in TRL too?

**Comment by muellerzr** [comment]
(Somewhat, currently trying to reverse engineer a few ways you did it, you guys would be *much* faster at it I imagine if you want to beat us to it ;) )

**Comment by danielhanchen** [comment]
:)
Wrote a detailed tweet about it: https://x.com/danielhanchen/status/1846235913443262891
Also Reddit post: https://www.reddit.com/r/LocalLLaMA/comments/1g4ego7/llm_training_bug_fixes_gradient_accumulation_was/
Blog post: https://unsloth.ai/blog/gradient
Also @shimmyshimmer is my brother!! :)

**Comment by muellerzr** [comment]
Just as a fair warning, this will not be an immediate nor quick fix, since essentially this means every single model's calculation is off when doing `output.loss`, and every single model will need a custom variation of CrossEntropy (and other valid loss funcs) if you do *not* calculate the loss by hand. 

We are working on figuring out the best solution.

**Comment by nahidalam** [comment]
@danielhanchen from the blog `The 2nd theory was there is in fact a bug in the loss calculation, which we find to be the case.` this bug is specifically for `CrossEntropy` loss calculation in HF `trl`? This will not be an issue if someone is using say [torch.nn.CrossEntropyLoss](https://pytorch.org/docs/stable/generated/torch.nn.CrossEntropyLoss.html) ?

**Comment by huseinzol05** [comment]
@muellerzr , i believe this only make sense padding based batch, for packing, there is no 0 / pad token in the batch, so avg cross entropy is consistent

**Comment by danielhanchen** [comment]
@nahidalam Unfortunately this is not a HF native issue. The way gradient accumulation has been originally done in many packages even those that use Pytorch directly accidentally missed considering ignored tokens. Using CE Loss directly does not solve the issue since mean reduction does not work, and sum will cause the loss to be scaled incorrectly.

@huseinzol05 Packing is also affected albeit less so since some people also do training on completions so it'll also make the loss incorrect.

@muellerzr If you guys need any help on anything, ping me!

**Comment by wongjingping** [comment]
Kudos @danielhanchen on the fix! Neat write-up as well! Back to the OP, I think the issue isn't with the `trl` library, but with the `transformers` library instead, because of how [SFTTrainer extends Trainer](https://github.com/huggingface/trl/blob/2ba3005d1c0a9ecec130108ef767d496b6d720cd/trl/trainer/sft_trainer.py#L67), how the loss is calculated in [Trainer's compute_loss](https://github.com/huggingface/transformers/blob/main/src/transformers/trainer.py#L3598-L3633), and how it is naively scaled by the number of steps [here](https://github.com/huggingface/transformers/blob/main/src/transformers/trainer.py#L3596).

**Comment by muellerzr** [comment]
Yes, ddp does we already have documented this + a fix is being put in

**Comment by qgallouedec** [comment]
Right @burtenshaw.
Closes by https://github.com/huggingface/transformers/pull/34198

**Comment by pminervini** [comment]
<img width="208" alt="Screenshot 2024-11-27 at 07 53 31" src="https://github.com/user-attachments/assets/077d0284-7980-407c-8c05-6db844120ed3">

Time to hit that "Close Issue" button @qgallouedec @burtenshaw! :) I thought the issue was open because of that!

**Comment by surprisedPikachu007** [comment]
> @huseinzol05 Packing is also affected albeit less so since some people also do training on completions so it'll also make the loss incorrect.

For language modeling task, will this be a problem even if all samples in a batch have the exact same sequence length?
